### ISSUE TITLE
color

### ISSUE THREAD

**Issue by pkra** [zorkow/speech-rule-engine#290]
I was wondering what your thoughts are on exposing color information. Right not it seems it is discarded which seems like a loss of information.

**Comment by zorkow** [comment]
Currently ignored but should probably change. Localisation will of course be an issue.

**Comment by pkra** [comment]
Here's a "wonderful" example from real life:

![image](https://github.com/Speech-Rule-Engine/speech-rule-engine/assets/426686/bd70b499-73e4-4bed-8ed8-0d43ec0013e5)

Note the text referencing the parts of a specific color. (The again, this is a rather extreme example failing WCAG in all sorts of ways.)

**Comment by dmca-glasgow** [comment]
+1 I came here to ask exactly the same question :)

**Comment by zorkow** [comment]
Thanks everyone of bringing this up again. In 2019 we did not have the mechanisms to deal with colour. That's different now, so we could make an effort get that in. Technically, it would not be difficult. The real question is: How would voicing look like?

Trivially one could announce the "start" and "end" of a colour.

But how about an examples like this 
$$\color{red}a+{\color{green}x}+b$$
or even this
$$\color{red}a{\color{green}x}^{2}$$
`Red a Green x Red squared`?

**Comment by dmca-glasgow** [comment]
I suggest "start <colour>" and "end <colour>" would be a good starting point.

Just based on your examples above, maybe additionally a shorthand for single characters, for example "red a".  

But "start red a end red" also doesn't seem too bad.

**Comment by pkra** [comment]
I noticed that SRE now adds color information. Adding start/end information seem unavoidable to me (though edge cases might get away with less information). I think and indication that the information identifies color is relevant, too - I saw `#007091 upper E 5` in a real world example which seems problematic without context.

**Comment by pkra** [comment]
On related note: is background color exposed? I couldn't create an example where it was.

**Comment by zorkow** [comment]
> I noticed that SRE now adds color information. Adding start/end information seem unavoidable to me (though edge cases might get away with less information). I think and indication that the information identifies color is relevant, too - I saw `#007091 upper E 5` in a real world example which seems problematic without context.

Yes, basic colour information is added, but currently only in MathSpeak English. The colour naming is simply derived from the attribute set by MathJax. So I assume that your example with the hex code stems from a `definecolor` command. MathJax passes that through as hex code and not via the custom colour name. While one could have a bespoke solution in MathJax, this would not solve the overall problem. A way around this would be to compute a colour name using a nearest neighbour function for a fixed set of named colors. I've done that in a different project before to compute meaningful aria labels, but that used the colorjs.io library which would be a overkill in this use case.

> On related note: is background color exposed? I couldn't create an example where it was.

No, not yet. It suffers from similar problems as foreground colour, where the tree nesting needs to be handled in speech. (See my examples above.)  It would be good to test how this could be done effectively, without too many `Start` and `End` elements.

Also, please feel free to send me more samples via email if you have any.
